src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+        }
+
+        private async Task<FrameHeader> ReadFrameAsync()
+        {
+            // Read frame header
+            await EnsureIncomingBytesAsync(FrameHeader.Size).ConfigureAwait(false);
+            FrameHeader frameHeader = FrameHeader.Read(_incomingBuffer.ReadBytes);
+            _incomingBuffer.Consume(FrameHeader.Size);
+
+            if (frameHeader.Length > FrameHeader.MaxLength)
+            {
+                throw new Http2ProtocolException(Http2ProtocolErrorCode.FrameSizeError);
+            }
+
+            // Read frame contents
+            await EnsureIncomingBytesAsync(frameHeader.Length).ConfigureAwait(false);
+
+            return frameHeader;
+        }
+
+        private async void ProcessIncomingFrames()
+        {
+            try
+            {
+                while (true)
+                {
+                    FrameHeader frameHeader = await ReadFrameAsync();
+
+                    switch (frameHeader.Type)
+                    {
+                        case FrameType.Headers:
+                            await ProcessHeadersFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.Data:
+                            ProcessDataFrame(frameHeader);
+                            break;
+
+                        case FrameType.Settings:
+                            await ProcessSettingsFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.Priority:
+                            ProcessPriorityFrame(frameHeader);
+                            break;
+
+                        case FrameType.Ping:
+                            await ProcessPingFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.WindowUpdate:
+                            ProcessWindowUpdateFrame(frameHeader);
+                            break;
+
+                        case FrameType.GoAway:          // TODO
+                        case FrameType.RstStream:       // TODO
+                            throw new NotImplementedException();
+
+                        case FrameType.PushPromise:     // Should not happen, since we turn this off in our settings
+                        case FrameType.Continuation:    // Should only be received while processing headers in ProcessHeadersFrame
+                        default:
+                            throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                // TODO: Shutdown handling 
+                // Send GOAWAY to the peer, unless they sent us a GOAWAY first
+                // Fail any pending requests
+                // Clear out reference to connection so that the next request will create a new connection
```

Comment:
Seems like even for the initial check in we should do something a bit more than just eat all of these exceptions, maybe even Fail Fast?  If this were to happen, it would seem currently it would manifest as hangs, right?

Reply:
Yeah, I agree. See comments above.